dev-manuals/backend/testing.md

```diff
@@ -1,4 +1,162 @@
+Otherwise, we would have to mock so much that the test would become unreadable and hard to maintain.
+
+We use [Mockito](https://site.mockito.org/) for mocking. It allows us to mock classes and interfaces and to verify
+that certain methods were called on the mock.
+
+### Naming conventions
+
+There are no strict naming conventions for unit tests. However, the test method should describe what is tested and
+what the expected result is. For example, if we test the delete method of the `UserService`, we could name the test
+method `testDeleteExistingUser()`.
+Other approaches put the expected result in the test method name, for example `deleteUser_UserExists_UserDeleted()`,
+but this often leads to very long test method names.
+
+We therefore prefer to put the expected result in the JavaDoc of the test method:
+The JavaDoc of the test method should describe what is tested and what the expected result is, after the given
+pattern: `Given <some context>, when <some action is taken>, then <some result is expected>`.
+
+### Example
+
+```java
+public class ChapterServiceTest {
+    /**
+     * Given a valid ChapterId
+     * When deleteChapter is called
+     * Then the UUID is returned and the delete method of the repository is called
+     */
+    @Test
+    public void testDeleteChapterSuccessful() {
+        // arrange test data
+        UUID testChapterId = UUID.randomUUID();
+
+        // mock repository
+        doNothing()
+                .when(chapterRepository).deleteById(any());
+        doReturn(true)
+                .when(chapterRepository).existsById(any());
+
+        // act
+        UUID deletedChapterId = chapterService.deleteChapter(testChapterId);
+
+        // assert
+        assertThat(deletedChapterId, is(testChapterId));
+
+        // verify that the repository was called
+        verify(chapterRepository).deleteById(testChapterId);
+    }
+}
+```
+
+## Integration/API testing
+
+As our microservices are rather small, we do API testing instead of what is usually referred to as integration testing.
+This means that we test the graphQL API of the microservices instead of interaction of classes in the microservice.
+Having integration tests in addition to unit tests and API tests would be redundant, as the API tests already test
+the interaction of the classes and the database.
+
+Of course, if it makes sense to have integration tests in addition in some cases, we can always add them.
+
+### What to test
+
+We test the graphQL API of the microservices. This means that we test the graphQL queries and mutations that are
+exposed by the microservice. As in unit tests, think of all special cases and exceptional paths that should be tested.
+
+If a functionality is tests in detail in a unit test, we do not have to test it again in that much detail in an API
+test. But in case of doubt, we prefer to have more tests than less.
+
```

Comment:
This is true for the Java implementation, however integration tests also test the database part. E.g. if the Java code does not set a non-nullable attribute, this could result in a database constraint violation. An error like that could be unnoticed in a unit test.

Reply:
But that would be tested by the API test, wouldn't it?

Reply:
That is true. So should I remove this section.

Reply:
Instead of removing this section, I think we should change it as follows: "Even if your unit tests test the complete functionality, an integration test might be necessary to test the database related aspects like database constraints."